engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -184,20 +212,56 @@ class WorkflowActor(workflowId: WorkflowId,
   when(WorkflowFailedState) { FSM.NullFunction }
   when(WorkflowSucceededState) { FSM.NullFunction }
 
-  onTransition {
-    case oldState -> terminalState if terminalState.terminal =>
-      log.info(s"$tag transition from $oldState to $terminalState: shutting down")
-      context.stop(self)
-    case fromState -> toState =>
-      log.info(s"$tag transitioning from $fromState to $toState")
-  }
-
   whenUnhandled {
+    case Event(MetadataPutFailed(action, error), _) =>
+      // Do something useful here??
+      log.warning(s"$tag Put failed for Metadata action $action : ${error.getMessage}")
+      stay
+    case Event(MetadataPutAcknowledgement(_), _) => stay()
     case Event(AbortWorkflowCommand, WorkflowActorData(Some(actor), _)) =>
       actor ! EngineLifecycleActorAbortCommand
       goto(AbortingWorkflowState)
     case unhandledMessage =>
       log.warning(s"$tag received an unhandled message $unhandledMessage in state $stateName")
       stay
   }
+
+  onTransition {
+    case fromState -> toState =>
+      log.info(s"$tag transitioning from $fromState to $toState")
+      // This updates the workflow status
+      pushCurrentStateToMetadataService(toState.workflowState)
+  }
+
+  onTransition {
+    case oldState -> terminalState if terminalState.terminal =>
+      log.info(s"$tag transition from $oldState to $terminalState: shutting down")
+      // Add the end time of the workflow in the MetadataService
+      val metadataEventMsg = MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.EndTime), MetadataValue(DateTime.now.toString))
+      serviceRegistryActor ! PutMetadataAction(metadataEventMsg)
```

Comment:
@Horneth With fire and forget, we no longer have futures which is in resonance with what Jeff wanted. However, we can no longer impose an order on the messages, i.e. earlier it would have received the failure message andThen send a workflow completion message to parent.  With that, if it had received a PoisonPill it would have done the appropriate action before being killed. 
We are not doing anything useful with the failures currently anyway, but it might be the case that the workflow completed and the parent sent a message to kill itself, and then it receives the Metadata failure message which will probably go to the dead letter mailbox (Sigh. Side effects of giving the metadata responsibility to this actor I suppose)